benchmark/e2e/test.js

```diff
@@ -0,0 +1,181 @@
+
+function delay (ms) {
+  return new Promise((resolve, reject) => {
+    setTimeout(resolve, ms)
+  })
+}
+
+function sh (cmd) {
+  return new Promise((resolve, reject) => {
+    console.log('>', cmd)
+    spawn(cmd, [], { stdio: 'inherit', shell: true })
+      .on('error', reject)
+      .on('close', resolve)
+  })
+}
+
+function forkProcess (file, options = {}) {
+  return new Promise((resolve, reject) => {
+    console.log(`> node ${options.execArgv ? options.execArgv.join(' ') + ' ' : ''}${file}`)
+    options.stdio = 'pipe'
+    const subProcess = fork(file, options)
+    console.log('>> PID', subProcess.pid)
+    subProcess.on('message', message => {
+      if (message.ready) {
+        resolve({ subProcess })
+      }
+    })
+  })
+}
+
+const statAsync = promisify(stat)
+async function exists (filename) {
+  try {
+    const stats = await statAsync(filename)
+    return stats.isDirectory() || stats.isFile()
+  } catch (e) {
+    return false
+  }
+}
+
+function get (url) {
+  return new Promise((resolve, reject) => {
+    _get(url, res => {
+      const chunks = []
+      res.on('data', d => chunks.push(d))
+      res.on('end', () => {
+        resolve(Buffer.concat(chunks).toString())
+      })
+    })
+  })
+}
+
+async function checkDb () {
+  console.log('# checking that db is populated')
+  cd('acmeair-nodejs')
+  const { subProcess } = await forkProcess('./app.js', {
+    execArgv: process.execArgv.concat(['--require', '../datadog.js'])
+  })
+
+  const customers = await get('http://localhost:9080/rest/api/config/countCustomers')
+
+  if (parseInt(customers) < 10000) {
+    console.log('# populating db')
+    await get('http://localhost:9080/rest/api/loader/load?numCustomers=10000')
+  }
+
+  subProcess.kill()
+  cd(__dirname)
+}
+
+async function ensureAppIsInstalled () {
+  cd(__dirname)
+  if (!(await exists(path.join(__dirname, 'acmeair-nodejs')))) {
+    await sh('git clone git@github.com:acmeair/acmeair-nodejs.git')
+  }
+  if (!(await exists(path.join(__dirname, 'acmeair-nodejs', 'node_modules')))) {
+    cd('acmeair-nodejs')
+    await sh('npm install')
+    cd(__dirname)
+  }
+}
+
+function runTest (url, duration) {
+  return autocannon({ url, duration })
+}
+
+async function testBoth (url, duration, prof) {
+  cd(__dirname)
+  const { subProcess: agentProcess } = await forkProcess('./fake-agent.js')
+  cd('acmeair-nodejs')
+  const execArgv = ['--require', '../datadog.js']
+  if (prof) {
+    execArgv.unshift('--prof')
+  }
+  const { subProcess: airProcess } = await forkProcess('./app.js', {
+    execArgv: execArgv.concat(process.execArgv),
+    env: Object.assign({}, process.env, { DD_ENABLE: '1' })
+  })
+
+  await delay(2000)
+
+  const resultWithTracer = await runTest(url, duration)
+
+  airProcess.kill()
+  agentProcess.kill()
+
+  const { subProcess: airProcess2 } = await forkProcess('./app.js', {
+    execArgv: execArgv.concat(process.execArgv)
+  })
+
+  const resultWithoutTracer = await runTest(url, duration)
```

Comment:
I *think* I know what you mean, but could you clarify anyway?

Reply:
I've just realized this is currently not setup in CircleCI, but I meant to parallelize in 6 jobs to avoid taking over a minute for the benchmark. Ideally it would be parallelized locally too, but if I understand correctly this is not really meant to be run locally.

Reply:
It is intended to be run locally.

If you parallelize it locally, the tests will interfere with each other.

Reply:
I'll add a TODO to make sure the different tests are runnable in isolation.